### ISSUE TITLE
Variable replacement not working in 'jobs' the same as 'deployment' jobs.

### ISSUE THREAD

**Issue by btnguyen2k** [microsoft/azure-pipelines-tasks#13350]
## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: bug

**Enter Task Name**: Docker@2

list here (V# not needed):  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

## Environment
- Server - Azure Pipelines or TFS on-premises? Azure Pipelines

- Agent - Hosted or Private: Microsoft Hosted Agent

## Issue Description

I have a simple task like this:

```
    - task: Docker@2
      displayName: Login to Docker Hub
      inputs:
        command: login
        containerRegistry: '$(dockerRegistry)'
```

If I define variable `dockerRegistry` within the pipeline file (inside `variables` block) then it works fine.
However, if I import it from Pipeline Library like this:

```
variables:
  - group: 'Release'
```

then it does not work, the error message was "There was a resource authorization issue: "The pipeline is not valid. Job Build_and_Publish: Step Docker1 input containerRegistry references service connection $(dockerRegistry) which could not be found. The service connection does not exist or has not been authorized for use. For authorization details, refer to https://aka.ms/yamlauthz."

I tried some simple echo and confirmed that `echo $(containerRegistry)` (and other variables in the same libary) works (i.e. the printed value is correct).

### Task logs

[Enable debug logging and please provide the zip file containing all the logs for a speedy resolution]

## Troubleshooting
Checkout how to troubleshoot failures and collect debug logs: https://docs.microsoft.com/en-us/vsts/build-release/actions/troubleshooting

### Error logs

[Insert error from the logs here for a quick overview]

**Comment by ds-ms** [comment]
Could you attach a snippet of your yamls?

**Comment by btnguyen2k** [comment]
Hi @ds-ms ,

Below is the working yaml, after workaround:

```
trigger:
  branches:
    include:
    - master
  paths:
    exclude:
    - docs/*

variables:
- group: 'Release'

stages:
- stage: Build
  displayName: Build and Publish Docker image
  pool:
    vmImage: 'ubuntu-latest'
  jobs:
  - job: Build
    displayName: Build and Publish Docker image
    # variables:
    # - group: 'Release'
    steps:
    - task: Docker@2
      displayName: Login to Docker Hub
      inputs:
        command: login
        containerRegistry: '$(dockerRegistry)'
    - task: Docker@2
      displayName: Build and Publish image
      inputs:
        command: buildAndPush
        dockerfile: '$(Build.SourcesDirectory)/Dockerfile'
        containerRegistry: '$(dockerRegistry)'
        repository: '$(dockerRepository)'
        tags: |
          $(Build.BuildId)
```

I have to move the `variables` to the top level to make it work. If I put `variables` block under stages or jobs I got the error `"There was a resource authorization issue: "The pipeline is not valid. Job Build_and_Publish: Step Docker1 input containerRegistry references service connection $(dockerRegistry) which could not be found. The service connection does not exist or has not been authorized for use. For authorization details, refer to https://aka.ms/yamlauthz."`

Best regards,

**Comment by ds-ms** [comment]
Overall the issue I see is, the variable being available when the replacement happens. If the dockerRegistry variable is available in context during the job orchestration, it would replace it with the value you're looking for. But in your case the value was not being parsed. So instead it was looking for a service connection named `$(dockerRegistry)` which didn't exist so it failed.

I'd like to understand more on what you meant by `if I import it from Pipeline Library` are you talking about a variable group here? If yes, then this is probably expected, replacing service connections is something which is not supported from variable groups. You'd probably notice this behaviour across all the other tasks, not just docker. I'll reconfirm this and get back to you.

**Comment by btnguyen2k** [comment]
@ds-ms, yes I was talking about variable group (which was created via pipelie's libary page).
If I put `variables` block at top level it works, if I put it inside `stages` or `jobs it does not.

**Comment by btnguyen2k** [comment]
One more thing, I have another task as the following:

```
- stage: Deploy
  displayName: Deploy App Service
  dependsOn: Build
  pool:
    vmImage: 'ubuntu-latest'
  jobs:
  - deployment: Deploy
    displayName: Deploy App Service from Container
    environment: dev
    variables:
    - group: Release
    strategy:
      runOnce:
        deploy:
          steps:
          - task: AzureWebAppContainer@1
            displayName: 'Azure Web App on Container Deploy'
            inputs:
              azureSubscription: '$(azureSubscription)'
              appName: '$(appName)'
              containers: '$(dockerRepository):$(Build.BuildId)'
```

Task `AzureWebAppContainer@1` works well with the `variables` block at `jobs` level.

**Comment by ds-ms** [comment]
That is weird. The orchestrator, I believe doesn't do any biased decisions like that. If it works for one task, it's expected to work with everything else. 

One difference I noticed is you're using Docker task in a regular job and AzureWebContainer task under a `deployment` stage. Could you validate if this is occurring for you under just `job` or `deployment job` under a `stage`?

**Comment by btnguyen2k** [comment]
Hi @ds-ms,

I think I kind of understand the error. I attached the error screenshot:
![image](https://user-images.githubusercontent.com/2164028/88891350-63cc3780-d286-11ea-9019-f6ba9b6c0610.png)

The error with Docker@2 task happens _before_ the pipeline was able to ever run any task. I think this check is a type of _pre-run_ check. And for some reasons at this _pre-run_ check step, the top level `variables` block is evaluated while the inner `variables` block is not.
I did another test: I hard-code the part `containerRegistry: 'actual-connector-name-here'` to make it run. And inside the job, I added another step to echo variables and I can confirm that all variables are populated correctly from variable group.

So in summary, I guess is that with `Docker@2` task this check happens before the pipeline runs, hence only the top-level `variables` block is evaluated.
Once the _pre-check_ step passes, the `variables` group works as expected (hence the task `AzureWebAppContainer@1` works without issue).

Hope it helps.

**Comment by ds-ms** [comment]
Like I mentioned it's probably not a task thing. 

I'm guessing it's a difference in behaviour between a `job` and `deployment job` type under a stage. If you're able to repro this issue under both the cases.. I'll engage the right team to help with this issue 😄

**Comment by btnguyen2k** [comment]
Hi @ds-ms 

Your guess is correct! I change the deployment job to normal job like this:

```
- stage: Deploy
  displayName: Deploy App Service
  pool:
    vmImage: 'ubuntu-latest'
  jobs:
  - job: Deploy
    displayName: Deploy App Service with Docker image
    variables:
    - group: 'Release'
    steps:
      - task: AzureWebAppContainer@1
        displayName: 'Azure Web App on Container Deploy'
        inputs:
          azureSubscription: '$(azureSubscription)'
          appName: '$(appName)'
          containers: '$(dockerRepository):$(Build.BuildId)'
```

And I got this error message:
![image](https://user-images.githubusercontent.com/2164028/88894090-b4459400-d28a-11ea-9b6f-2c7052f91b59.png)

So the issue now is why the `variables` group block is not consistently evaluated? I expect it to work without putting the `variables` block at top-level.

**Comment by ds-ms** [comment]
@zachariahcox there seems to be an issue with parsing variables from a variable group in job layer. Is this an expected behaviour?

**Comment by blumk** [comment]
@ds-ms @zachariahcox 
This bug does affects other tasks as well, like `AzureRmWebAppDeployment@4`.

Variables don't get resolved (default is empty string) and task fails with error message:
`##[error]Error: Input required: WebAppName`

I'm using `deployment job`, trying to migrate to YAML CD.

**Comment by Zachery2008** [comment]
I found the the same error, and I wish someone help to solve it. Thanks

**Comment by Zachery2008** [comment]
I found a solution. You can save the acr connection as a variable in a variable group. And if you want to choose different variable groups depending on some conditions, you can do it at the top level. So the `containerRegistry` will treat it correctly. 
For example,
```
trigger:
- master

variables:
  # Agent VM image name
  - name: vmImageName
    value: ubuntu-latest
  - name: imageRepository
    value: xxx
  - name: dockerfilePath
    value: xxx
  - name: buildContextPath
    value: xxx
  - ${{ if eq(variables['Build.SourceBranch'], 'refs/heads/master') }}:
    - group: abc123
  - ${{ if eq(variables['Build.SourceBranch'], 'refs/heads/prod') }}:
    - group: xyz456

```

**Comment by dhindrik** [comment]
I have a similar issue:


```
jobs:
- job: BuildSE${{ upper(parameters.env) }}
  variables:
    - group: myapp-ssh-key-${{ lower(parameters.env) }}
    - group: myapp-docker-${{ lower(parameters.env) }}
    - group: myapp-frontend-${{ lower(parameters.env) }}
    - group: myapp-terraform-${{ lower(parameters.env) }}
    - group: myapp-terraform-backend-${{ lower(parameters.env) }}
  steps:
  - template: container-template.yml
    parameters:
        env: ${{ upper(parameters.env) }}
        country: SE
```
container-template.yml
```
parameters:
    - name: country
      type: string
      default: ''
    
    - name: env
      type: string
      default: ''  

    

steps:
- task: Docker@2
  displayName: 'Build container for ${{ upper(parameters.country) }} - ${{ upper(parameters.env) }}'
  inputs:
    containerRegistry: $(containerRegistry)
    repository: 'myapp/${{ lower(parameters.country) }}'
    command: 'build'
    Dockerfile: '**/Dockerfile'
    arguments: --build-arg APP_ENV=${{ lower(parameters.env) }} --build-arg COUNTRY_CODE=${{ lower(parameters.country) }} --build-arg BACKEND_URL=$(apim-public-api) --build-arg SUBSCRIPTION_KEY=$(apim-subscription-key) --build-arg APPLICATION_INSIGHTS_KEY=$(ApplicationInsightsKey)
    
- task: Docker@2
  condition: not(startswith(variables['Build.SourceBranch'], 'refs/pull'))
  displayName: Push container for ${{ upper(parameters.country) }} - ${{ upper(parameters.env) }}
  inputs:
    containerRegistry: $(containerRegistry)
    repository: 'myapp/${{ lower(parameters.country) }}'
    command: 'push'
    Dockerfile: '**/Dockerfile'
```

There was a resource authorization issue: "The pipeline is not valid. Job BuildSEDEV: Step Docker1 input containerRegistry references service connection $(containerRegistry) which could not be found. The service connection does not exist or has not been authorized for use. For authorization details, refer to https://aka.ms/yamlauthz.,Job BuildSEDEV: Step Docker2 input containerRegistry references service connection $(containerRegistry) which could not be found. The service connection does not exist or has not been authorized for use. For authorization details, refer to https://aka.ms/yamlauthz."

**Comment by sonnehansen** [comment]
Thank you all for the investigation :)

An identical issue occurs for variables from a template in a `jobs` section (and works for `deployment` type):

```
jobs:
- job: build_and_push_image  
  variables:
    - template: variables.yml@repo
  steps:
...
   - task: Docker@1
      displayName: 'Build Docker Image'
      inputs:
        containerregistrytype: 'Azure Container Registry'
        azureSubscriptionEndpoint: $(ACRSubscriptionEndpoint)
        azureContainerRegistry: $(ACR)
        command: 'Build an image'
        dockerFile: 'Dockerfile'
        imageName: '$(ACR)/$(ImageName):$(Build.BuildNumber)'
```    

- Inside `jobs` it does not resolve (but `imageName` does because it is outside initial validation):
![image](https://user-images.githubusercontent.com/29034546/104287274-82319280-54b6-11eb-9467-96d476c48aab.png)
- Moving `template: variables.yml@repo` to the top level it works, as reported by @btnguyen2k.

@ds-ms: I has been a while since there was any activity here. Can you confirm if MS will fix the issue because it is a bug, or we must settle for the (top level) work-around?

**Comment by BartNetJS** [comment]
Same issue here: variabele ServiceConnectionK8SDEFAULT from a template file is not substituted in the pre run check

```
- stage: DeployTST
  condition: and(succeeded())
  displayName: Deploy TEST
  dependsOn: DeployDev
  pool:
    vmImage: $(LinuxVmImageName)
  variables:
    - template: pipelines/configurations/mfiles/for-mfiles-pipelines-tst.yml@sharedTemplatesRepository
    - template: pipelines/configurations/for-all-pipelines-tst.yml@sharedTemplatesRepository
  jobs:
  - job: Deploy_AAD_Identity_Validation_TEST
    displayName: Deploy AAD Identity Validation on TEST
    steps:
    - template: pipelines/aadpodidentity/aadpodidentity-deploy-template.yml@sharedTemplatesRepository
      parameters:
        kubernetesServiceEndpoint: $(ServiceConnectionK8SDEFAULT)
        kubernetesServiceConnection: $(ServiceConnection-K8S)
        AADPodValidationName: $(AADPodValidationName)
        namespace: $(Namespace)
```

and the template file:
```
variables:
  # Azure DevOps Service Connection
  ServiceConnectionK8SDEFAULT: BDOBE-T-AKS-bdobe-ns-default
```
![image](https://user-images.githubusercontent.com/4865355/107018574-e63b3400-67a0-11eb-897b-2952cdf1f0c6.png)

**Comment by lbouriez** [comment]
Any workaround ?
The ${{ variables.XXX }} doesn't work in our case :/

**Comment by Zachery2008** [comment]
> Any workaround ?
> 
> The ${{ variables.XXX }} doesn't work in our case :/

Did you check out my above comment? You can save the variables in the group and import it.

**Comment by lbouriez** [comment]
> > Any workaround ?
> > 
> > The ${{ variables.XXX }} doesn't work in our case :/
> 
> Did you check out my above comment? You can save the variables in the group and import it.

Yes, i used this one for the moment but my stage is inside a template and i wanted it to have the variables too... Not to load the variables group from the top of the pipeline :(

**Comment by adilsonmenechini** [comment]
```yaml
    variables:
      ${{ if eq(variables['Build.SourceBranchName'], 'develop') }}: 
        dockerRegistry: "Nexus-nonprod"
        dockerRegistryDomain: "docker-nonprod.url"
      ${{ if or(eq(variables['Build.SourceBranchName'], 'master'), eq(variables['Build.SourceBranchName'], 'release'), eq(variables['Build.SourceBranchName'], 'release/*')) }}: 
        dockerRegistry: "Nexus-prod"
        dockerRegistryDomain: "docker.url"
```

**Comment by richatcosta** [comment]
This still appears to be an issue - and as per the comment from @sonnehansen it is an issue when wanting to use a (dynamically-selected) variable group in a template (or variable template etc):

`release-pipeline.yml` file:
```
stages:
- stage: UAT
  jobs:
  - template: build-template.yml@pipelineTemplates
    parameters:
      variable_group: my-uat-vg
```

`build-template.yml` file:
```
parameters:
- name: variable_group
  type: string    

jobs:
  variables:
  - group: ${{ parameters.variable_group }}
  - name: something_else
    value: 'value here'

  strategy:
    runOnce:
      deploy:
        steps:
...
```

Attempting to refer to a variable in the variable group as part of the steps results in the errors posted earlier eg "references $(foo) which was not found"

@ds-ms @zachariahcox it would be good to get an update from Microsoft on this - if any fix is planned etc. Or if not, what the recommended workaround is, and official documentation around this if possible.

**Comment by snelis** [comment]
Did anyone find a solution or workaround for this ?

**Comment by spasal** [comment]
If I remember correctly it worked for me when using the syntax `${{ variables['NameOfVariable.WithDotNotation'] }}`

Please see [Runtime expression syntax](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#runtime-expression-syntax), `$(var)` is processed runtime before a task executes, but the service connection is checked before that.

Good luck!

**Comment by ForteUnited** [comment]
How is this still an issue? omg

**Comment by ForteUnited** [comment]
So what is the solution here?

**Comment by pavelmateju** [comment]
> If I remember correctly it worked for me when using the syntax `${{ variables['NameOfVariable.WithDotNotation'] }}`
> 
> Please see [Runtime expression syntax](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#runtime-expression-syntax), `$(var)` is processed runtime before a task executes, but the service connection is checked before that.
> 
> Sometimes moving the variable group top level is not ideal, as the connection might be different per stage.
> 
> Good luck!

Thank you . That helped

**Comment by idontusenumbers** [comment]
Though I was able to use the `${{ variables['NameOfVariable.WithDotNotation'] }}` workaround for `azureSubscription` input on a `AzureFunctionApp`,  the same workaround doesn't work with `job.environment`. I can't get any of the syntaxes to work there.

The workaround syntax isn't documented on https://learn.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch

**Comment by channeladam** [comment]
Same issue.  Variable groups listed at the job deployment level are not working.
When I look at the pipeline logs, the `azure-pipelines-expanded.yaml` file has those job deployment level variables deleted - so I assume that is why the variable expansion doesn't happen.

Hey Microsoft-related people, @davidstaheli @bryanmacfarlane @prawalagarwal @mvvsubbu or anyone, do you think you could kindly somehow get some engagement from Microsoft on this almost 3 year bug that contradicts the documentation on how variables work at different levels? 

It is a pretty fundamental issue...

https://learn.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/variables?view=azure-pipelines#remarks

**Comment by channeladam** [comment]
Ok this is odd... hours later I re-ran the pipeline and it worked.

Now the deployment level variables are no longer being deleted in the `azure-pipelines-expanded.yaml` file.

Nothing really changed on my end...

**Comment by mvvsubbu** [comment]
@channeladam I am not working on Azure DevOps anymore. I will let other team members help you with this issue if it is still continuing.

**Comment by dom93dd** [comment]
I figured out a way how to use different variable groups for different stages, even if the variable groups share the same variable names (with different values). I still don't understand why Microsoft hasn't fixed this...

**main.yaml:**
```
variables:
- group: VARIABLE_GROUP_A
- group: VARIABLE_GROUP_B

stages:
- stage: BUILD_DEPLOY_GROUP_A
  displayName: 'Build/Deploy A'
  variables:
  - group: VARIABLE_GROUP_A

jobs:
  - template: deploy-environment.yml
    parameters:
      environment: GROUP_A

- stage: BUILD_DEPLOY_GROUP_B
  displayName: 'Build/Deploy B'
  variables:
  - group: VARIABLE_GROUP_B

jobs:
  - template: deploy-environment.yml
    parameters:
      environment: GROUP_B
```

**deploy-environment.yaml:**
```
parameters:
- name: environment
  type: string
  default: ""

jobs:
- deployment: Deploy
  dependsOn: Build
  pool:
    vmImage: 'ubuntu-latest'
  environment: ${{ parameters.environment }}

  strategy:
    runOnce:
      deploy:
        steps:
...
Now you are able to use the variables from selected variable group from main.yaml.
```

Somehow different variable groups only work with attribute "- deployment" within jobs. Don't ask me why. As a workaround I used "- deployment" also for my build steps. 

Microsoft should offer different variable groups for different stages...

Hopefully this helps somebody.
